clients/client/src/main/java/org/projectnessie/client/NessieClientBuilder.java

```diff
@@ -0,0 +1,95 @@
+/*
+ * Copyright (C) 2020 Dremio
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.projectnessie.client;
+
+import java.net.URI;
+import java.util.function.Function;
+import org.projectnessie.client.NessieClient.AuthType;
+
+/**
+ * {@link NessieClient} builder interface.
+ *
+ * @param <IMPL> concrete nessie-client builder type
+ */
+public interface NessieClientBuilder<IMPL extends NessieClientBuilder<IMPL>> {
```

Comment:
we discussed a protocol version and another parameter here also right?

Is this the only (future) way to get a Nessie client?

Reply:
Yes, this should become the only way to build+get a Nessie client. The current `NessieClient.Builder` shall go away.

Once this one and a follow-up to add the "API version" + "API invocation builder pattern" are in, we can nuke `NessieClient` + the current `TreeApi` + `ConfigApi` + `ContentsApi` interfaces. But it's a long(er) way to get there.